src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -398,7 +398,13 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
 
                     // SPEC: When this is used in a primary-expression within an instance method or instance accessor
                     // SPEC: of a struct, it is classified as a variable. 
-                    if (!thisref.Type.IsValueType || RequiresReturnableReference(valueKind))
+
+                    // "this" is only readable if we are not in a struct. RValueOnly is checked at the beginning of this method
+                    if (!thisref.Type.IsValueType ||
+                        //"this" is only readable in members of readonly structs, uless weare in a constructor
+                        (thisref.Type.IsReadOnly && (this.ContainingMember() as MethodSymbol)?.MethodKind != MethodKind.Constructor) ||
+                        //"this" is not returnable by reference in a struct
```

Comment:
I kind of thought that commenting on different conditions separately would be more readable. Perhaps not.

---
In reply to: [118026950](https://github.com/dotnet/roslyn/pull/19713#discussion_r118026950) [](ancestors = 118026950)